Control/Concurrent/MVar/Strict.hs

```diff
@@ -63,22 +66,27 @@ import Control.DeepSeq
 --     fairness properties of abstractions built using 'MVar's.
 --
 putMVar  :: NFData a => MVar a -> a -> IO ()
-#ifndef __HADDOCK__
-putMVar (MVar mvar#) !x = rnf x `seq` IO $ \ s# -> -- strict!
-    case putMVar# mvar# x s# of
-        s2# -> (# s2#, () #)
-#endif
+putMVar !mv !x = rnf x `seq` MV.putMVar mv x
 
 -- | A non-blocking version of 'putMVar'.  The 'tryPutMVar' function
 -- attempts to put the value @a@ into the 'MVar', returning 'True' if
 -- it was successful, or 'False' otherwise.
 --
 tryPutMVar  :: NFData a => MVar a -> a -> IO Bool
-#ifndef __HADDOCK__
-tryPutMVar (MVar mvar#) !x = IO $ \ s# -> -- strict!
-    case tryPutMVar# mvar# x s# of
-        (# s, 0# #) -> (# s, False #)
-        (# s, _  #) -> (# s, True #)
+tryPutMVar !mv !x = rnf x `seq` MV.tryPutMVar mv x
+
+#if !MIN_VERSION_base(4,7,0)
+-- |A non-blocking version of 'readMVar'.  The 'tryReadMVar' function
+-- returns immediately, with 'Nothing' if the 'MVar' was empty, or
+-- @'Just' a@ if the 'MVar' was full with contents @a@.
+tryReadMVar :: MVar a -> IO (Maybe a)
+-- This is a best-effort compatibility shim for really old GHC versions.
+-- It's not really what you'd call *right*.
+tryReadMVar !m = uninterruptibleMask $ \_ -> do
```

Comment:
Actually, `mask` should be enough, since nothing in here is interruptible. An earlier draft used `putMVar`. I'm not sure which version makes the least nonsense. But ultimately, it doesn't really matter much, since this is a half-broken shim. (`readMVar` itself should be considered half-broken for earlier `base` versions, since it used to be non-atomic and potentially disruptive.)